### ISSUE TITLE
`None` gradients with bijectors

### ISSUE THREAD

**Issue by ColCarroll** [tensorflow/probability#1190]
Bijector cacheing causes `None` gradients:

```python
with tf.GradientTape() as tape:
  d = tfd.MultivariateNormalTriL(scale_tril=tf.eye(2))
  x = d.sample()  
  tape.watch(x)
  g = tape.gradient(d.log_prob(x), x)

g is None  # True
tfp.math.value_and_gradient(d.log_prob, x)[1] is None  # True
```

The basic cause of this is that `bijector.inverse(bijector.forward(x))` returns `x` from the cache, and the gradient of this will be `None`. 

Currently, this can be worked around in the above example by breaking the cache. For example, replacing `x = d.sample()` with any of 

```python
x = d.sample() + 0
x = tf.identity(d.sample())
x = tf.stop_gradient(d.sample())
```

**Comment by ColCarroll** [comment]
This was the cause of #1182
